pkg/policy/l4.go

```diff
@@ -1084,13 +1111,280 @@ func addL4Filter(policyCtx PolicyContext,
+	Diff(t *testing.T, expectedMap L4PolicyMap) string
+	Len() int
+}
+
+// NewL4PolicyMap creates an new L4PolicMap.
+func NewL4PolicyMap() L4PolicyMap {
+	return &l4PolicyMap{
+		namedPortMap: make(map[string]*L4Filter),
+		rangePortMap: bitlpm.NewUintTrie[uint32, map[portProtoKey]*L4Filter](),
+	}
+}
+
+// InitL4PolicyMap creates an new L4PolicMap, with an initial
+// set of values. The initMap argument does not support ranges.
+func InitL4PolicyMap(initMap map[string]*L4Filter) L4PolicyMap {
+	l4M := &l4PolicyMap{
+		namedPortMap: make(map[string]*L4Filter),
+		rangePortMap: bitlpm.NewUintTrie[uint32, map[portProtoKey]*L4Filter](),
+	}
+	for k, v := range initMap {
+		portProtoSlice := strings.Split(k, "/")
+		if len(portProtoSlice) < 2 {
+			continue
+		}
+		l4M.Upsert(portProtoSlice[0], 0, portProtoSlice[1], v)
+	}
+	return l4M
+}
+
+type portProtoKey struct {
+	port, endPort uint16
+	proto         uint8
+}
+
+// l4PolicyMap is the implementation of L4PolicyMap
+type l4PolicyMap struct {
+	namedPortMap map[string]*L4Filter
+	// rangePortMap has to keep a map of L4Filters rather than
+	// a single L4Filter reference so that the l4PolicyMap does
+	// not merge L4Filter that are not the same port range, but
+	// share an overlapping range in the trie.
+	rangePortMap *bitlpm.UintTrie[uint32, map[portProtoKey]*L4Filter]
+	// rangeMapLen must be tracked separately from
+	// rangePortMap as L4Filters are split up when
+	// the port range is long enough. This does not
+	// represent additional entries only the vagaries
+	// of indexing by range.
+	rangeMapLen int
+}
+
+func parsePortProtocol(port, protocol string) (uint16, uint8) {
+	// These string values have been validated many times
+	// over at this point.
+	prt, _ := strconv.ParseUint(port, 10, 16)
+	proto, _ := u8proto.ParseProtocol(protocol)
+	return uint16(prt), uint8(proto)
+}
+
+func makePolicyMapKey(port, mask uint16, proto uint8) uint32 {
+	return (uint32(proto) << 16) | uint32(port&mask)
+}
+
+// Upsert L4Filter adds an L4Filter indexed by protocol/port-endPort.
+func (l4M *l4PolicyMap) Upsert(port string, endPort uint16, protocol string, l4 *L4Filter) {
+	if iana.IsSvcName(port) {
+		l4M.namedPortMap[port+"/"+protocol] = l4
+		return
+	}
+
+	portU, protoU := parsePortProtocol(port, protocol)
+	ppK := portProtoKey{
+		port:    portU,
+		endPort: endPort,
+		proto:   protoU,
+	}
+	var upsertHappened bool
+	for _, mp := range PortRangeToMaskedPorts(portU, endPort) {
+		k := makePolicyMapKey(mp.port, mp.mask, protoU)
+		prefix := 32 - uint(bits.TrailingZeros16(mp.mask))
+		portProtoMap, ok := l4M.rangePortMap.ExactLookup(prefix, k)
+		if !ok {
+			portProtoMap = make(map[portProtoKey]*L4Filter)
+			l4M.rangePortMap.Upsert(prefix, k, portProtoMap)
+		}
+		if !upsertHappened {
+			if _, ok := portProtoMap[ppK]; !ok {
+				l4M.rangeMapLen += 1
+				upsertHappened = true
+			}
+		}
+		portProtoMap[ppK] = l4
+	}
+}
+
+// Delete delets an L4Filter from the index by protocol/port-endPort
```

Comment:
nit: Stutter, spelling
```suggestion
// Delete an L4Filter from the index by protocol/port-endPort
```